src/infrasys/system.py

```diff
@@ -597,6 +597,22 @@ def get_component_types(self) -> Iterable[Type[Component]]:
         """
         return self._component_mgr.get_types()
 
+    def list_parent_components(
```

Comment:
I'm open to a better name. `list_connected_components` or `list_attached_components` would be misleading because this won't return buses connected to generators. `list_containing_components`, `list_components_connected_upstream`.

The function could be `list_connected_components(self, upstream=True, component_type=None)` and then allow it to work either direction.